build_modules/lib/src/meta_module.dart

```diff
@@ -0,0 +1,433 @@
+import 'dart:async';
+import 'dart:collection';
+import 'dart:math';
+
+import 'package:analyzer/analyzer.dart';
+import 'package:build/build.dart';
+import 'package:path/path.dart' as p;
+import 'package:json_annotation/json_annotation.dart';
+
+import 'modules.dart';
+
+part 'meta_module.g.dart';
+
+/// Returns the top level directory in [uri].
+///
+/// Throws an [ArgumentError] if [uri] is just a filename with no directory.
+String _topLevelDir(String uri) {
+  var parts = p.url.split(p.url.normalize(uri));
+  String error;
+  if (parts.length == 1) {
+    error = 'The uri `$uri` does not contain a directory.';
+  } else if (parts.first == '..') {
+    error = 'The uri `$uri` reaches outside the root directory.';
+  }
+  if (error != null) {
+    throw new ArgumentError(
+        'Cannot compute top level dir for path `$uri`. $error');
+  }
+  return parts.first;
+}
+
+/// Convert [importUri] found in [source] to an [AssetId], handling both
+/// `package:` imports and relative imports.
+///
+/// Returns null for `dart:` uris since they cannot be referenced properly
+/// by an `AssetId`.
+///
+/// Throws an [ArgumentError] if an [AssetId] can otherwise not be created. This
+/// might happen if:
+///
+/// * [importUri] is absolute but has a scheme other than `dart:` or `package:`.
+/// * [importUri] is a relative path that reaches outside of the current top
+///   level directory of a package (relative import from `web` to `lib` for
+///   instance).
+AssetId _importUriToAssetId(AssetId source, String importUri) {
```

Comment:
You can use `AssetId.parse` instead of this, https://github.com/dart-lang/build/blob/master/build/lib/src/asset/id.dart#L54

Reply:
Done.